beacon-chain/sync/data_column_sidecars.go

```diff
@@ -0,0 +1,854 @@
+// only if the indices are the same all blocks and if the blocks are contiguous.
+// (Missing blocks or blocks without commitments do count as contiguous)
+// If one of this condition is not met, returns nil.
+func buildByRangeRequests(
+	slotByRoot map[[fieldparams.RootLength]byte]primitives.Slot,
+	slotsWithCommitments map[primitives.Slot]bool,
+	indicesByRoot map[[fieldparams.RootLength]byte]map[uint64]bool,
+	batchSize uint64,
+) ([]*ethpb.DataColumnSidecarsByRangeRequest, error) {
+	if len(indicesByRoot) == 0 {
+		return nil, nil
+	}
+
+	var reference map[uint64]bool
+	slots := make([]primitives.Slot, 0, len(slotByRoot))
+	for root, indices := range indicesByRoot {
+		if reference == nil {
+			reference = indices
+		}
+
+		if !compareIndices(reference, indices) {
+			return nil, nil
+		}
+
+		slot, ok := slotByRoot[root]
+		if !ok {
+			return nil, errors.Errorf("slot not found for block root %#x", root)
+		}
+
+		slots = append(slots, slot)
+	}
+
+	slices.Sort(slots)
+
+	for i := 1; i < len(slots); i++ {
+		previous, current := slots[i-1], slots[i]
+		if current == previous+1 {
+			continue
+		}
+
+		for j := previous + 1; j < current; j++ {
+			if slotsWithCommitments[j] {
+				return nil, nil
+			}
+		}
+	}
+
+	columns := sortedSliceFromMap(reference)
+	startSlot, endSlot := slots[0], slots[len(slots)-1]
+	totalCount := uint64(endSlot - startSlot + 1)
+
+	requests := make([]*ethpb.DataColumnSidecarsByRangeRequest, 0, totalCount/batchSize)
+	for start := startSlot; start <= endSlot; start += primitives.Slot(batchSize) {
+		end := min(start+primitives.Slot(batchSize)-1, endSlot)
+		request := &ethpb.DataColumnSidecarsByRangeRequest{
+			StartSlot: start,
+			Count:     uint64(end - start + 1),
+			Columns:   columns,
+		}
+
+		requests = append(requests, request)
+	}
+
+	return requests, nil
+}
+
+// buildByRootRequest constructs a by root request from the given indices.
+func buildByRootRequest(indicesByRoot map[[fieldparams.RootLength]byte]map[uint64]bool) p2ptypes.DataColumnsByRootIdentifiers {
+	identifiers := make(p2ptypes.DataColumnsByRootIdentifiers, 0, len(indicesByRoot))
+	for root, indices := range indicesByRoot {
+		identifier := &eth.DataColumnsByRootIdentifier{
+			BlockRoot: root[:],
+			Columns:   sortedSliceFromMap(indices),
+		}
+		identifiers = append(identifiers, identifier)
+	}
+	return identifiers
+}
+
+// verifyDataColumnSidecarsByPeer verifies the received data column sidecars.
+// If at least one sidecar from a peer is invalid, the peer is downscored and
+// all its sidecars are rejected. (Sidecars from other peers are still accepted.)
+func verifyDataColumnSidecarsByPeer(
+	p2p prysmP2P.P2P,
+	newVerifier verification.NewDataColumnsVerifier,
+	roDataColumnsByPeer map[goPeer.ID][]blocks.RODataColumn,
+) ([]blocks.VerifiedRODataColumn, error) {
+	// First optimistically verify all received data columns in a single batch.
+	count := 0
+	for _, columns := range roDataColumnsByPeer {
+		count += len(columns)
```

Comment:
count seemed to be incremented but i don't see it used anywhere for verification